graphstorm-processing/tests/test_dist_hard_negative_transformation.py

```diff
@@ -0,0 +1,107 @@
+def test_hard_negative_example_list(spark: SparkSession, check_df_schema, tmp_path):
+    # Input Data DataFrame
+    data = [
+        ("mark", "doctor", ["scientist"]),
+        ("john", "scientist", ["engineer", "nurse"]),
+        ("tara", "engineer", ["nurse", "doctor", "scientist"]),
+        ("jen", "nurse", ["doctor"]),
+    ]
+    columns = ["src_type", "dst_type", "hard_negative"]
+    input_df = spark.createDataFrame(data, schema=columns)
+
+    # Mapping DataFrame
+    mapping_data = [
+        ("doctor", 0),
+        ("scientist", 1),
+        ("engineer", 2),
+        ("nurse", 3),
+    ]
+    mapping_column = [NODE_MAPPING_STR, NODE_MAPPING_INT]
+    mapping_df = spark.createDataFrame(mapping_data, schema=mapping_column)
+    mapping_df.repartition(1).write.parquet(f"{tmp_path}/raw_id_mappings/dst_type/parquet")
+    edge_mapping_dict = {
+        "edge_type": "src_type:relation:dst_type",
+        "mapping_path": f"{tmp_path}/raw_id_mappings/",
+        "format_name": "parquet",
+    }
+    hard_negative_transformation = DistHardEdgeNegativeTransformation(
+        ["hard_negative"], spark=spark, edge_mapping_dict=edge_mapping_dict, separator=None
+    )
+    output_df = hard_negative_transformation.apply(input_df)
+    check_df_schema(output_df)
+    output_data = output_df.collect()
+
+    expected_output = [[1, -1, -1, -1], [2, 3, -1, -1], [3, 0, 1, -1], [0, -1, -1, -1]]
+
+    for idx, row in enumerate(output_data):
+        np.testing.assert_almost_equal(
+            row[0], expected_output[idx], decimal=3, err_msg=f"Row {idx} is not equal"
+        )
+
+
+def test_hard_negative_example_str(spark: SparkSession, check_df_schema, tmp_path):
+    # Input Data DataFrame
+    data = [
+        ("mark", "doctor", "scientist"),
+        ("john", "scientist", "engineer;nurse"),
+        ("tara", "engineer", "nurse;doctor;scientist"),
+        ("jen", "nurse", "doctor"),
+    ]
+    columns = ["src_type", "dst_type", "hard_negative"]
+    input_df = spark.createDataFrame(data, schema=columns)
+
+    # Mapping DataFrame
+    mapping_data = [
+        ("doctor", 0),
+        ("scientist", 1),
+        ("engineer", 2),
+        ("nurse", 3),
+    ]
+    mapping_column = [NODE_MAPPING_STR, NODE_MAPPING_INT]
+    mapping_df = spark.createDataFrame(mapping_data, schema=mapping_column)
+    mapping_df.repartition(1).write.parquet(f"{tmp_path}/raw_id_mappings/dst_type/parquet")
+    edge_mapping_dict = {
+        "edge_type": "src_type:relation:dst_type",
+        "mapping_path": f"{tmp_path}/raw_id_mappings/",
+        "format_name": "parquet",
+    }
+    hard_negative_transformation = DistHardEdgeNegativeTransformation(
+        ["hard_negative"], spark=spark, edge_mapping_dict=edge_mapping_dict, separator=";"
+    )
+    output_df = hard_negative_transformation.apply(input_df)
+    check_df_schema(output_df)
+    output_data = output_df.collect()
+
+    expected_output = [[1, -1, -1, -1], [2, 3, -1, -1], [3, 0, 1, -1], [0, -1, -1, -1]]
```

Comment:
Same question.

Reply:
Same as above, 4 kinds of nodes for the dst node type.